Feature/5 google map api

pozafly opened the pull request:
## 주제

- Google Map API 연동
- Google Place API로 autocomplete 컴포넌트 추가

## 작업 내용

- 다이어리 우측에 지도 나타나도록 추가
- Lexical Editor 모듈화
- Lexical Editor에 Google Place API 연동

pozafly committed:
feat: 페이지에 스크롤을 없애기 위해 uiUtils 작성

pozafly committed:
feat: diary map 컴포넌트 생성 및 레이아웃 수정

pozafly committed:
chore: add @googlemaps library

pozafly committed:
feat: diary map 컴포넌트에 map api 연동 및 hooks 설정

pozafly committed:
fix: SSR 환경이 아닌 곳에 존재하는 alert 분기 처리

pozafly committed:
refactor: toolbar plugin의 Block, Link 플러그인 분리 및 스타일 수정

pozafly committed:
refactor: toolbar plugin의 나머지 모두 모듈화

pozafly committed:
refactor: editor 드롭다운 로직 uiUtils로 분리

pozafly committed:
feat: google autocomplete 컴포넌트의 css 추가

pozafly committed:
feat: google autocomplete context 추가

pozafly committed:
feat: google autocomplete api 연동 및 Map Toolbar 컴포넌트 추가

f-lab-james commented on src/constants/map.ts:
```diff
@@ -0,0 +1,3 @@
+// 송파구
+export const POTISION = { lat: 37.51175556, lng: 127.1079306 } as const;
```
```tsx
export const DEFAULT_POTISION = Object.freeze({ lat: 37.51175556, lng: 127.1079306 })
```

f-lab-james commented on src/core/utils/uiUtils.ts:
```diff
@@ -0,0 +1,35 @@
+  editor: HTMLElement,
+  rect: DOMRect | undefined | null
+) => {
+  if (rect === null) {
```
```ts
if (!rect) {
    editor.style.opacity = '0';
    editor.style.top = '-1000px';
    editor.style.left = '-1000px';
} else {
    editor.style.opacity = '1';
    editor.style.top = `${
        rect.top + rect.height + window.pageYOffset + 10
    }px`;
    const left =
      rect.left +
      window.pageXOffset -
      editor.offsetWidth / 2 +
      rect.width / 2;
      editor.style.left = `${left >= 0 ? left : 10}px`;
}
```

f-lab-james commented on src/features/diary/components/content/DiaryEditor/DiaryEditor.tsx:
```diff
@@ -49,59 +49,49 @@ export default function DiaryEditor({ document, metaData, diaryId }: Props) {
+        onError(error: Error) {
+          throw error;
+        },
+        editorState: document ? JSON.stringify(document) : null,
```
요것은 취향인 것 같네요

```ts
(document && JSON.stringify(document)) ?? null
```

f-lab-james commented on src/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/index.tsx:
```diff
@@ -0,0 +1,159 @@
+import UnderlineToolbar from '@/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/UnderlineToolbar.tsx';
+import StrikethroughToolbar from '@/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/StrikethroughToolbar.tsx';
+import DirectionToolbar from '@/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/DirectionToolbar.tsx';
+import MapToolbar from '@/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/mapToolbar/index.tsx';
```
```ts
// @/features/diary/components/content/DiaryEditor/plugins
export * from '~~./DirectionToolbar.tsx'
export * from '~~./ItalicToolbar.tsx'
export * from '~~./StrikethroughToolbar.tsx'
```
re-export 한 후

```ts
import { DirectionToolbar, ItalicToolbar, StrikethroughToolbar } from '@/features/diary/components/content/DiaryEditor/plugins'
```

f-lab-james commented on src/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/index.tsx:
```diff
@@ -0,0 +1,159 @@
+import DirectionToolbar from '@/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/DirectionToolbar.tsx';
+import MapToolbar from '@/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/mapToolbar/index.tsx';
+
+const LowPriority = 1;
```
LowPriority -> LOW_PRIORITY

f-lab-james commented on src/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/BlockToolbar.tsx:
```diff
@@ -0,0 +1,228 @@
+    const toolbar = toolbarRef.current;
+    const dropDown = dropDownRef.current;
+
+    if (toolbar !== null && dropDown !== null) {
```
```ts
if (toolbar && dropDown) {
}
```

f-lab-james commented on src/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/LinkToolbar.tsx:
```diff
@@ -0,0 +1,198 @@
+      let rect: DOMRect | undefined;
+      if (nativeSelection?.anchorNode === rootElement) {
+        let inner = rootElement;
+        while (inner.firstElementChild != null) {
```
=이 하나 빠졌네요

f-lab-james commented on src/features/diary/components/content/DiaryEditor/plugins/ToolbarPlugin/subToolbars/UndoToolbar.tsx:
```diff
@@ -0,0 +1,36 @@
+        <i className="format undo" />
+      </button>
+      <button
+        disabled={!canRedo}
```
can대신 disabled로 관리하는게 더 이해하기 쉽지 않을까요?

f-lab-james commented on src/features/diary/hooks/useMapLoad.ts:
```diff
@@ -0,0 +1,42 @@
+        zoom: ZOOM,
+      });
+    });
+  }, [mapRef]);
```
ref가 생기는 타이밍을 잡기 위해 useState에 ref를 넣는것으로 보이는데요!

useLayoutEffect를 써도 ref가 처음에 null인지 확인만 한 번 해주세요!

pozafly replied:
넵 useLayoutEffect도 null로 나타나서 그대로 진행해야할 것 같습니다 ㅠ

pozafly committed:
refactor: constants object 형태를 Object.freeze로 안전하게 보관하도록 수정

pozafly committed:
chore: add body-scroll-lock libaray

pozafly committed:
feat: uiUtils의 scroll 관련 로직 걷어내고 body-scroll-lock 라이브러리 적용

pozafly committed:
style: 자잘한 코드 스타일 변경

1. util 함수 분기문 간단하게 수정
2. css 스타일에 필요없는 옵션 제거
3. 글로벌 스타일에 font(antialiased)관련 스타일 추가
4. Lexical editor에 toolbar plugin을 re-export 형태로 변경

pozafly committed:
refactor: 변수명 변경 canUndo -> diabledUndo, canRedo -> disabledRedo

pozafly committed:
fix: type 정의되지 않은 부분 정의
